lite2/client.go

```diff
@@ -0,0 +1,374 @@
+// algorithm.
+//
+// If the trusted header is more recent than one here, an error is returned.
+func (c *Client) VerifyHeader(newHeader *types.SignedHeader, newVals *types.ValidatorSet, now time.Time) error {
+	if c.trustedHeader.Height >= newHeader.Height {
+		return errors.Errorf("height #%d is already trusted (last: #%d)", newHeader.Height, c.trustedHeader.Height)
+	}
+
+	var err error
+	switch c.mode {
+	case sequential:
+		err = c.sequence(newHeader, newVals, now)
+	case skipping:
+		err = c.bisection(c.trustedHeader, c.trustedNextVals, newHeader, newVals, now)
+	}
+	if err != nil {
+		return err
+	}
+
+	// Update trusted header and vals.
+	nextVals, err := c.primary.ValidatorSet(newHeader.Height + 1)
+	if err != nil {
+		return err
+	}
+	return c.updateTrustedHeaderAndVals(newHeader, nextVals)
+}
+
+func (c *Client) sequence(newHeader *types.SignedHeader, newVals *types.ValidatorSet, now time.Time) error {
+	// 1) Verify any intermediate headers.
+	var (
+		interimHeader *types.SignedHeader
+		nextVals      *types.ValidatorSet
+		err           error
+	)
+	for height := c.trustedHeader.Height + 1; height < newHeader.Height; height++ {
+		interimHeader, err = c.primary.SignedHeader(height)
+		if err != nil {
+			return errors.Wrapf(err, "failed to obtain the header #%d", height)
+		}
+
+		err = Verify(c.chainID, c.trustedHeader, c.trustedNextVals, interimHeader, c.trustedNextVals, c.trustingPeriod, now, c.trustLevel)
+		if err != nil {
+			return errors.Wrapf(err, "failed to verify the header #%d", height)
+		}
+
+		// Update trusted header and vals.
+		if height == newHeader.Height-1 {
+			nextVals = newVals
+		} else {
+			nextVals, err = c.primary.ValidatorSet(height + 1)
+			if err != nil {
+				return errors.Wrapf(err, "failed to obtain the vals #%d", height+1)
+			}
+		}
+		if !bytes.Equal(interimHeader.NextValidatorsHash, nextVals.Hash()) {
+			return errors.Errorf("expected next validator's hash %X, but got %X (height #%d)",
+				interimHeader.NextValidatorsHash,
+				nextVals.Hash(),
+				height)
+		}
+		err = c.updateTrustedHeaderAndVals(interimHeader, nextVals)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update trusted state #%d", height)
+		}
+	}
+
+	// 2) Verify the new header.
+	return Verify(c.chainID, c.trustedHeader, c.trustedNextVals, newHeader, newVals, c.trustingPeriod, now, c.trustLevel)
+}
+
+func (c *Client) bisection(
+	lastHeader *types.SignedHeader,
+	lastVals *types.ValidatorSet,
+	newHeader *types.SignedHeader,
+	newVals *types.ValidatorSet,
+	now time.Time) error {
+
+	err := Verify(c.chainID, lastHeader, lastVals, newHeader, newVals, c.trustingPeriod, now, c.trustLevel)
+	switch err.(type) {
+	case nil:
+		return nil
+	case types.ErrTooMuchChange:
+		// continue bisection
+	default:
+		return errors.Wrapf(err, "failed to verify the header #%d ", newHeader.Height)
+	}
+
+	if newHeader.Height == c.trustedHeader.Height+1 {
+		// TODO: submit evidence here
```

Comment:
Evidence of what?

Reply:
misbehavior? when a primary provider sends us an invalid header

Reply:
But this would be `TooMuchChange` between two subsequent headers, right?

I'm still not sure when this case would actually arise. Header `h + 1` is either valid with respect to `h` or it is not, there is no notion of `TooMuchChange` in that case, since we have the `NextValSetHash`.

Reply:
> there is no notion of TooMuchChange in that case

isn't there? what if new validator set (suppose `NextValSetHash` matches and we trust it) did not sign the commit (i.e. < 2/3 signed). `VerifyCommit` will return `ErrTooMuchChange` in that case. I agree this is confusing and we should probably return different error, say `ErrNotEnoughValsSigned` in `VerifyCommit`. What do you think?

Reply:
Ah I see - yes, I think that should be a different error then.

Reply:
We should probably not use term evidence in this context. This is error due to malformed header but not necessarily evidence of slashable behaviour. Evidence comes from cross-checking verified headers and it should probably be done as part of failure detection module.